mivideo/mivideo_extract.py

```diff
@@ -0,0 +1,115 @@
+# -*- coding: utf-8 -*-
+import logging
+import os
+import time
+from typing import Union, Sequence, Dict
+
+import pandas as pd
+from google.cloud import bigquery
+from google.oauth2 import service_account
+from sqlalchemy.engine import ResultProxy
+
+import mivideo.queries as queries
+from db.db_creator import DBCreator
+from environ import ENV, CONFIG_DIR
+from run_jobs import ValidDataSourceName
+
+logger = logging.getLogger(__name__)
+
+SHAPE_ROWS: int = 0  # Index of row count in DataFrame.shape() array
+
+
+class MiVideoExtract(object):
+    udpKeyFilePath: str
+    credentials: service_account.Credentials
+    udpDb: bigquery.Client
+    appDb: DBCreator
+
+    def __init__(self):
+        self.udpKeyFilePath = os.path.join(CONFIG_DIR, ENV.get('mivideo', {}).get(
+            'service_account_json_filename'))
+        logger.debug(f'udpKeyFilePath: "{self.udpKeyFilePath}"')
+
+        self.credentials = service_account.Credentials.from_service_account_file(
+            self.udpKeyFilePath,
+            scopes=['https://www.googleapis.com/auth/cloud-platform'],
+        )
+
+        self.udpDb = bigquery.Client(
+            credentials=self.credentials,
+            project=self.credentials.project_id,
+        )
+
+        logger.debug(f'Connected to BigQuery project: "{self.udpDb.project}"')
+
+        DB_PARAMS = ENV['INVENTORY_DB']
+        APPEND_TABLE_NAMES = ENV.get('APPEND_TABLE_NAMES', ['job_run'])
+
+        self.appDb = DBCreator(DB_PARAMS, APPEND_TABLE_NAMES)
+        self.appDb.set_up()
+
+    def _readTableLastTime(self, tableName: str, tableColumnName: str) -> Union[str, None]:
+        lastTime: str = None
+
+        try:
+            sql: str = f'select max(t.{tableColumnName}) from {tableName} t'
```

Comment:
Since you're really just using hard-coded values in `run` and this is only used once, it might just be better to use them here in the string, to avoid any SQL injection warnings.

Reply:
If you need to use this for other processes you plan to add, not sure we have a great solution to the SQL injection warnings. I've been trying to deal with this too, and using parameters for database and table names with MySQL seems to be a non-starter; see [this StackOverflow question](https://stackoverflow.com/questions/40249590/sqlalchemy-error-when-adding-parameter-to-string-sql-query)

Reply:
If it's completely necessary, I think we should just make sure we're validating inputs where appropriate.

Reply:
Yeah, it's used only internally in one place for now, but there will be a second MiVideo table added soon.  Actually, the other table was to be part of this PR.  However, its data isn't being used by the dashboard now.  So, I moved it to a separate issue so I could get this PR merged faster without it.

Where are you seeing the SQL injection warnings?  I don't see them when my code runs.  Are you using a different code quality checker or is Codacy telling us about this?

Reply:
Yes, the SQL injection warnings are coming from Codacy. Their scan doesn't seem very fine-grained or consistent, but they seem to flag SQL-like strings that have concatenation or interpolation.

Reply:
I've addressed most of the Codacy issues, I think.  I wish I knew of a faster way to get Codacy to check the code again locally, without needing to check changes into the repo.

WRT its issues about SQL injection, from what I looked up about SQLAlchemy and prevention of injection, it looks like the suggestion is to use its model features.  I'm not sure whether we want to get into that at this time.  I'm not opposed to it, but it seems like it will take an extra chunk of time to do it.

Reply:
Agreed, I also looked at that, but we'd essentially have to adopt the whole ORM I think for the project, which wouldn't be bad but would mean we'd have to remove `yoyo` and learn a new technology. At which point, why aren't are we just using Django, which we know?